### PR TITLE
[C, iOS, Android] LayoutCompression

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
### Description of Change ###

Enable Layout Compression for iOS and Android.

**NOTE: this feature is still in beta, and opt-in only**

Layout Compression allows the user to avoid creating renderers for Layouts (i.e. StackLayout, ContentView, ...).

Pros:
 - less renderers to create, manage and dispose
 - less UI elements on screen

Cons:
 - compressed layouts can't have BG colors, gesture recognizers
 - no transformation
 - ...

Sample:
```
<?xml version="1.0" encoding="UTF-8"?>
<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
	xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
	x:Class="Xamarin.Forms.Controls.CompressedLayout"
	Padding="0,20,0,0">
	<StackLayout Padding="6" CompressedLayout.IsHeadless="true">
		<Label Text="First"/>
		<StackLayout Padding="6" CompressedLayout.IsHeadless="true">
			<Label Text="Second"/>
			<ContentView Padding="6" CompressedLayout.IsHeadless="true" BackgroundColor="Pink" >
				<Label x:Name="this" Text="Third"/>
			</ContentView>
			<Label Text="Fourth"/>
		</StackLayout>
		<Label Text="Fifth"/>
	</StackLayout>
</ContentPage>
```
Enabling LayoutCompression for this layout only creates a Renderer for the page, and one for each label. The BackgroundColor property on the inner ContentView is ignored.

### Bugs Fixed ###

/

### API Changes ###

Added:
```
public static readonly BindableProperty CompressedLayout.IsHeadlessProperty;
```

### Behavioral Changes ###

/

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Platform.Android/VisualElementPackager.cs [hunk 1]
```diff
@@ -18,11 +18,16 @@ public class VisualElementPackager : IDisposable
 
 		IVisualElementRenderer _renderer;
 
-		public VisualElementPackager(IVisualElementRenderer renderer)
+		VisualElement _element;
+
+		IElementController ElementController => _element;// _renderer.Element as IElementController;
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/VisualElementPackager.cs]: Remove commented code

File: Xamarin.Forms.Core/Layout.cs [hunk 2]
```diff
@@ -237,6 +237,13 @@ protected void UpdateChildrenLayout()
 			double w = Math.Max(0, width - Padding.HorizontalThickness);
 			double h = Math.Max(0, height - Padding.VerticalThickness);
 
+			var isHeadless = CompressedLayout.GetIsHeadless(this);
+			var headlessOffset = CompressedLayout.GetHeadlessOffset(this);
+			for (var i = 0; i < LogicalChildrenInternal.Count; i++) {
```
Comment (samhouts) [inline Xamarin.Forms.Core/Layout.cs]: Braces on their own lines. You're in Core territory, now! muwahaha

File: Xamarin.Forms.Platform.Android/Renderers/LocalizedDigitsKeyListener.cs [hunk 3]
```diff
@@ -34,7 +34,9 @@ public static NumberKeyListener Create(InputTypes inputTypes)
 			if ((inputTypes & InputTypes.NumberFlagDecimal) == 0)
 			{
 				// If decimal isn't allowed, we can just use the Android version
+#pragma warning disable 0618 // Find issues with Android API usage
```
Comment (hartez) [inline Xamarin.Forms.Platform.Android/Renderers/LocalizedDigitsKeyListener.cs]: Which part of the next line is giving an obsolete warning?

File: Xamarin.Forms.Platform.Android/VisualElementPackager.cs (hunk 1 above)
Comment (hartez) [inline Xamarin.Forms.Platform.Android/VisualElementPackager.cs]: Can probably remove this comment.

File: Xamarin.Forms.Platform.iOS/RendererPool.cs [hunk 4]
```diff
@@ -108,10 +108,14 @@ void FillChildrenWithRenderers(VisualElement element)
 				var child = logicalChild as VisualElement;
 				if (child != null)
 				{
+					if (CompressedLayout.GetIsHeadless(child)) {
+						child.IsPlatformEnabled = true;
+						FillChildrenWithRenderers(child);
+					} else {
```
Comment (hartez) [inline Xamarin.Forms.Platform.iOS/RendererPool.cs]: Missing a tab on the next line.

File: Xamarin.Forms.Platform.Android/VisualElementPackager.cs [hunk 5]
```diff
@@ -67,44 +70,49 @@ protected virtual void Dispose(bool disposing)
...
 		}
 
 		void AddChild(VisualElement view, IVisualElementRenderer oldRenderer = null, RendererPool pool = null, bool sameChildren = false)
 		{
 			Performance.Start();
 
-			if (_childViews == null)
-				_childViews = new List<IVisualElementRenderer>();
+			if (CompressedLayout.GetIsHeadless(view)) {
+				var _packager = new VisualElementPackager(_renderer, view);
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/VisualElementPackager.cs]: the underscore in the variable name is a little confusing here, I think.

File: Xamarin.Forms.Platform.iOS/VisualElementPackager.cs [hunk 6]
```diff
@@ -13,16 +13,20 @@ public class VisualElementPackager : IDisposable
 
 		bool _isDisposed;
 
-		IElementController ElementController => Renderer.Element as IElementController;
+		IElementController ElementController => _element;// Renderer.Element as IElementController;
```
Comment (hartez) [inline Xamarin.Forms.Platform.iOS/VisualElementPackager.cs]: Can probably remove this comment.

**Comment by adrianknight89** [comment]
> compressed layouts can't have BG colors, gesture recognizers
> no transformation

Are these limitations permanent or to be fixed in the future?

`<ContentView Padding="6" CompressedLayout.IsHeadless="true" BackgroundColor="Pink" >`

Should there be a compile-time / run-time error for this instead of silent fail of the bgcolor set? The reason I ask this is I don't like it when people pollute their visual tree with redundant stuff. If bgcolor has no use there, then it shouldn't be there?

**Comment by StephaneDelcroix** [comment]
@adrianknight89 at this point we have no plan for supporting BgColor on headless layouts, nor do we have plans for validating that case

File: Xamarin.Forms.Platform.Android/Renderers/LocalizedDigitsKeyListener.cs (hunk 3 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Platform.Android/Renderers/LocalizedDigitsKeyListener.cs]: On X.A 7.5.0.15 (Alpha channel at this time): 'DigitsKeyListener.GetInstance(bool, bool)' is obsolete

```
[Register ("getInstance", "(ZZ)Landroid/text/method/DigitsKeyListener;", ""), Obsolete ("deprecated")]
public unsafe static DigitsKeyListener GetInstance (bool sign, bool @decimal);
```

File: Xamarin.Forms.Platform.Android/VisualElementPackager.cs (hunk 1 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Platform.Android/VisualElementPackager.cs]: fixed

File: Xamarin.Forms.Platform.iOS/RendererPool.cs (hunk 4 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Platform.iOS/RendererPool.cs]: fixed

File: Xamarin.Forms.Platform.iOS/VisualElementPackager.cs (hunk 6 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Platform.iOS/VisualElementPackager.cs]: fixed

File: Xamarin.Forms.Platform.Android/VisualElementPackager.cs (hunk 1 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Platform.Android/VisualElementPackager.cs]: fixed

File: Xamarin.Forms.Platform.Android/VisualElementPackager.cs (hunk 5 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Platform.Android/VisualElementPackager.cs]: fixed. at some point that variable was a field...

File: Xamarin.Forms.Core/Layout.cs (hunk 2 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/Layout.cs]: don't scratch that itch! :)

fixed by removing the braces.

File: Xamarin.Forms.Platform.Android/Renderers/LocalizedDigitsKeyListener.cs (hunk 3 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Platform.Android/Renderers/LocalizedDigitsKeyListener.cs]: fixed in master

**Comment by alexsorokoletov** [comment]
@StephaneDelcroix thank you for this PR and this feature. Wondering, where can one read about limitations and why they are there. Maybe it's just me can't find the information

**Comment by bentmar** [comment]
Can we please have some more information on how to use this?

**Comment by davidortinau** [comment]
@bentmar I'll have a blog out shortly with more samples, and then docs to come. Really any layout that you don't need a background on, doesn't transform, and doesn't accept gestures is a candidate. So just add `CompressedLayout.IsHeadless="true"` to it. I suspect we may discover some other cons as we test further. So far the reduction of renderers created is looking good, especially if you have lots of layout nesting. In my page I saw nearly a 50% reduction, though granted it was a poorly constructed page in many respects to begin with.

**Comment by alexsorokoletov** [comment]
https://blog.xamarin.com/3-big-things-explore-xamarin-forms-2-5-0-pre-release/

**Comment by BlueRaja** [comment]
Seems like this should be on by default, and only disabled when there's a background/transform/gesture/explicitly disabled?

**Pull request merged**